### DISCUSSION TITLE
SSHFS, NFS, NBD and persistence

### DISCUSSION THREAD

**Discussion by Soylor** [ltsp/ltsp#789] [Q&A] [answered]
Hi, i've moved to new ltsp, but have some questiones about persistence.

Previously LTSP i used NBD.
Can i boot from iPXE using NBD (How to?)
When login mounting NFS /home, then user homes are persistent.
When login using SSHFS, user homes are not persistent (Previous ltsp was persistent). 
Can i keep NOT using NFS and get persistent home user dirs?

**Comment by alkisg** [comment]
NBD had various issues. After years of trying to work around them, we switched to NFS which is much more stable.
It's possible to use NBD to boot LTSP, but it's not automated at all, you'd need to configure it manually. I'm not aware of anyone having posted instructions for this.

> When login using SSHFS, user homes are not persistent (Previous ltsp was persistent).

This means you didn't set up something correctly. SSHFS is persistent. Login to a client using the default NFS root + SSHFS home setup, type `mount` and paste the output here, so that we see which parts are wrong.

**Comment by Soylor** [comment]
![image](https://user-images.githubusercontent.com/9199475/216181100-c3790fa6-ab27-47e9-a07d-4a701aaa3b82.png)

**Reply by Soylor** [reply]
That's the mount output after login

**Comment by alkisg** [comment]
Copy/paste the output instead of uploading a screenshot.
Now the bottom part is clipped.

**Reply by Soylor** [reply]
````
siatec@EXTERNO3:~$ mount
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=973880k,nr_inodes=243470,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=202268k,mode=755,inode64)
tmpfs on /run/initramfs/ltsp type tmpfs (rw,relatime,mode=755,inode64)
/root/images/jammy.img on /run/initramfs/ltsp/0/looproot type squashfs (ro,relatime,errors=continue)
/run/initramfs/ltsp on / type overlay (rw,relatime,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work,xino=off)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
bpf on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=17747)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
none on /run/credentials/systemd-sysusers.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
tmpfs on /run/qemu type tmpfs (rw,nosuid,nodev,relatime,mode=755,inode64)
172.19.46.11:/home on /home type nfs (rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,nolock,proto=tcp,port=2049,timeo=600,retrans=3,sec=sys,local_lock=all,addr=172.19.46.11)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=202264k,nr_inodes=50566,mode=700,uid=1000,gid=1000,inode64)
gvfsd-fuse on /run/user/1000/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
portal on /run/user/1000/doc type fuse.portal (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
siatec@EXTERNO3:~$
```

**Comment by alkisg** [comment]
In this paste you've configured and you are using NFS home.
Weren't you asking about SSHFS?

**Reply by Soylor** [reply]
oh, yes, i disable and paste again. Sorry

**Comment by Soylor** [comment]
´´´
siatec@EXTERNO3:~$ mount
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=973880k,nr_inodes=243470,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=202268k,mode=755,inode64)
tmpfs on /run/initramfs/ltsp type tmpfs (rw,relatime,mode=755,inode64)
/root/images/jammy.img on /run/initramfs/ltsp/0/looproot type squashfs (ro,relatime,errors=continue)
/run/initramfs/ltsp on / type overlay (rw,relatime,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work,xino=off)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
bpf on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=17768)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
none on /run/credentials/systemd-sysusers.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
tmpfs on /run/qemu type tmpfs (rw,nosuid,nodev,relatime,mode=755,inode64)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=202264k,nr_inodes=50566,mode=700,uid=1000,gid=1000,inode64)
gvfsd-fuse on /run/user/1000/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
portal on /run/user/1000/doc type fuse.portal (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
siatec@EXTERNO3:~$

**Reply by Soylor** [reply]
```
siatec@EXTERNO3:~$ mount
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=973880k,nr_inodes=243470,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=202268k,mode=755,inode64)
tmpfs on /run/initramfs/ltsp type tmpfs (rw,relatime,mode=755,inode64)
/root/images/jammy.img on /run/initramfs/ltsp/0/looproot type squashfs (ro,relatime,errors=continue)
/run/initramfs/ltsp on / type overlay (rw,relatime,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work,xino=off)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
bpf on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=17768)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
none on /run/credentials/systemd-sysusers.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
tmpfs on /run/qemu type tmpfs (rw,nosuid,nodev,relatime,mode=755,inode64)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=202264k,nr_inodes=50566,mode=700,uid=1000,gid=1000,inode64)
gvfsd-fuse on /run/user/1000/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
portal on /run/user/1000/doc type fuse.portal (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
siatec@EXTERNO3:~$
```

**Comment by alkisg** [comment]
home isn't mounted in that paste.
Do you have sshfs installed?

**Reply by Soylor** [reply]
It's just a ltsp image created and booted
Server users can log in without problem
Can i run any comand to test sshfs or any other ting?

**Reply by Soylor** [reply]
when i do **ls /home**, then server home is showed

**Comment by alkisg** [comment, marked as answer]
Type `sshfs` in the client terminal to see if it's there or not.

If not, then add it to your ltsp image using `sudo apt install sshfs`.

**Reply by Soylor** [reply]
No, it's not instaled in image.
Yo're rigt, after instaling everithing works fine.
I followed LTSP instructions for creating debootstrap chroot and then ltsp image. ([CHROOT guide])(https://github.com/ltsp/ltsp/wiki/chroots)
No instructions about installing **sshfs** are there.
I think that can be modified the guide to include LTSP instructions about ltsp sshfs requiremnts.

For not rebuilding image i tested modifying **ltsp.conf**:
```
 POST_INIT_02="apt-get update && apt-get install sshfs
```
and then **ltsp initrd**
